File: typescript/src/well-known.ts
import { importSPKI, type KeyLike } from 'jose';
import type { WellKnownPayload, IVpsRouter } from './types.js';

const CACHE_TTL_MS = 60 * 60 * 1000; // 1 hour

interface KeyCache {
  keys: Map<string, KeyLike>;
  fetchedAt: number;
}

/** Fetches and caches the public-key list at `${vpsBase}/.well-known/fdkey.json`.
 *  Goes through the VpsRouter so the request lands on the same IP currently
 *  serving production traffic (and uses the same dispatcher / SNI / cert
 *  handling — just like challenge/submit calls). */
export class WellKnownClient {
  private cache: KeyCache | null = null;

  constructor(private readonly router: IVpsRouter) {}

  async getKey(kid: string): Promise<KeyLike | null> {
    if (this.cache && Date.now() - this.cache.fetchedAt < CACHE_TTL_MS) {
      const k = this.cache.keys.get(kid);
      if (k) return k;
      // kid not in cache — may have just rotated, refetch once
    }
    await this.refresh();
    return this.cache!.keys.get(kid) ?? null;
  }

  private async refresh(): Promise<void> {
    const target = await this.router.getTarget();
    const init: RequestInit & { dispatcher?: unknown } = {
      signal: AbortSignal.timeout(5000),
    };
    if (target.dispatcher) init.dispatcher = target.dispatcher;
    const res = await fetch(
      `${target.url}/.well-known/fdkey.json`,
      init as RequestInit
    );
    if (!res.ok) throw new Error(`fdkey: well-known fetch failed ${res.status}`);
    const payload = (await res.json()) as WellKnownPayload;
    const keys = new Map<string, KeyLike>();
    for (const k of payload.keys) {
      const key = await importSPKI(k.public_key_pem, k.alg);
      keys.set(k.kid, key);
    }
    this.cache = { keys, fetchedAt: Date.now() };
  }
}

File: rust/src/well_known.rs
//! `${vps_base}/.well-known/fdkey.json` cache. Mirrors the TS SDK's
//! `well-known.ts` — `HashMap<kid, DecodingKey>` cached for 1 hour,
//! refreshes on unknown kid (mid-rotation handling).
//!
//! Concurrency model: a `tokio::sync::RwLock` for the cache state plus a
//! separate `Mutex<()>` to serialize refreshes. When N concurrent callers
//! arrive on a cold cache, exactly ONE actually fetches; the others wait
//! on the refresh-mutex and read the freshly-populated cache when it
//! releases. No duplicate network calls, no races where one fetch
//! overwrites a newer one.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

use jsonwebtoken::DecodingKey;
use serde::Deserialize;
use tokio::sync::{Mutex, RwLock};

use crate::types::FdkeyError;

const CACHE_TTL: Duration = Duration::from_secs(60 * 60);
const FETCH_TIMEOUT: Duration = Duration::from_secs(5);

#[derive(Deserialize)]
struct WellKnownPayload {
    keys: Vec<WellKnownKey>,
}

#[derive(Deserialize)]
struct WellKnownKey {
    alg: String,
    kid: String,
    public_key_pem: String,
}

#[derive(Clone)]
pub struct WellKnownClient {
    vps_base: String,
    cache: Arc<RwLock<CacheState>>,
    refresh_lock: Arc<Mutex<()>>,
    /// Single reqwest::Client shared across refreshes. Reusing the client
    /// keeps the underlying connection pool warm — relevant when the
    /// server runs through a TTL-driven refresh storm shortly after
    /// startup. Building a fresh Client per call would discard the pool
    /// and pay a TLS handshake every time.
    http: reqwest::Client,
}

struct CacheState {
    keys: HashMap<String, DecodingKey>,
    fetched_at: Option<Instant>,
}

impl WellKnownClient {
    /// Build a new well-known client. Fallible because `reqwest::Client::
    /// builder().build()` can fail in stripped-down container images
    /// (missing CA bundle, locked-down rustls feature combos, etc.).
    /// With our default `rustls-tls` configuration it's infallible in
    /// practice, but we propagate the error rather than panicking so an
    /// integrator running in a non-standard environment gets a clean
    /// `FdkeyError` instead of a panic at startup.
    pub fn new(vps_base: impl Into<String>) -> Result<Self, FdkeyError> {
        let http = reqwest::Client::builder()
            .timeout(FETCH_TIMEOUT)
            .build()?;
        Ok(Self {
            vps_base: vps_base.into(),
            cache: Arc::new(RwLock::new(CacheState {
                keys: HashMap::new(),
                fetched_at: None,
            })),
            refresh_lock: Arc::new(Mutex::new(())),
            http,
        })
    }

    /// Returns a clone of the matching `DecodingKey` (or None if the kid
    /// isn't in the well-known). Triggers a refresh on cache miss / expiry.
    /// Concurrent callers coalesce on a single refresh — the second one
    /// through the refresh-lock sees the freshly-populated cache and skips
    /// the fetch.
    pub async fn get_key(&self, kid: &str) -> Result<Option<DecodingKey>, FdkeyError> {
        if let Some(key) = self.try_read_cache(kid).await {
            return Ok(Some(key));
        }
        // Cache miss / expiry. Take the refresh-lock to serialize.
        let _refresh = self.refresh_lock.lock().await;
        // Re-check: another task may have just refreshed while we waited.
        if let Some(key) = self.try_read_cache(kid).await {
            return Ok(Some(key));
        }
        self.refresh().await?;
        let cache = self.cache.read().await;
        Ok(cache.keys.get(kid).cloned())
    }

    async fn try_read_cache(&self, kid: &str) -> Option<DecodingKey> {
        let cache = self.cache.read().await;
        let fetched_at = cache.fetched_at?;
        if fetched_at.elapsed() >= CACHE_TTL {
            return None;
        }
        cache.keys.get(kid).cloned()
    }

    async fn refresh(&self) -> Result<(), FdkeyError> {
        let url = format!("{}/.well-known/fdkey.json", self.vps_base);
        let res = self.http.get(&url).send().await?;
        if !res.status().is_success() {
            let status = res.status().as_u16();
            let body = res.text().await.unwrap_or_default();
            return Err(FdkeyError::VpsHttp { status, body });
        }
        let payload: WellKnownPayload = res.json().await?;
        let mut keys = HashMap::new();
        for k in payload.keys {
            // FDKEY only issues EdDSA today; pinning is enforced in jwt_verify.
            let _ = k.alg;
            match DecodingKey::from_ed_pem(k.public_key_pem.as_bytes()) {
                Ok(key) => {
                    keys.insert(k.kid, key);
                }
                Err(err) => {
                    return Err(FdkeyError::Other(format!(
                        "fdkey: failed to import key {}: {err}",
                        k.kid
                    )));
                }
            }
        }
        let mut cache = self.cache.write().await;
        cache.keys = keys;
        cache.fetched_at = Some(Instant::now());
        Ok(())
    }
}
